htlcswitch/switch.go

```diff
@@ -867,60 +907,28 @@ func (s *Switch) handleLocalResponse(pkt *htlcPacket) {
 		return
 	}
 
-	// Locate the pending payment to notify the application that this
-	// payment has failed. If one is not found, it likely means the daemon
-	// has been restarted since sending the payment.
-	payment := s.findPayment(pkt.incomingHTLCID)
+	paymentID := pkt.incomingHTLCID
 
-	var (
-		preimage   [32]byte
-		paymentErr error
-	)
-
-	switch htlc := pkt.htlc.(type) {
-
-	// We've received a settle update which means we can finalize the user
-	// payment and return successful response.
-	case *lnwire.UpdateFulfillHTLC:
-		// Persistently mark that a payment to this payment hash
-		// succeeded. This will prevent us from ever making another
-		// payment to this hash.
-		err := s.control.Success(pkt.circuit.PaymentHash)
-		if err != nil && err != ErrPaymentAlreadyCompleted {
-			log.Warnf("Unable to mark completed payment %x: %v",
-				pkt.circuit.PaymentHash, err)
-			return
-		}
-
-		preimage = htlc.PaymentPreimage
-
-	// We've received a fail update which means we can finalize the user
-	// payment and return fail response.
-	case *lnwire.UpdateFailHTLC:
-		// Persistently mark that a payment to this payment hash failed.
-		// This will permit us to make another attempt at a successful
-		// payment.
-		err := s.control.Fail(pkt.circuit.PaymentHash)
-		if err != nil && err != ErrPaymentAlreadyCompleted {
-			log.Warnf("Unable to ground payment %x: %v",
-				pkt.circuit.PaymentHash, err)
-			return
+	// A pending payment for this paymentID should already exist for us to
+	// be able to notify the caller about the result. If it does not exist,
+	// we have probably restarted, and the caller hasn't yet attempted to
+	// resend the HTLC. To handle this case we create the pending payment
+	// now and populate it with the result, such that the caller can
+	// immediately parse it when a resend happens.
+	s.pendingMutex.Lock()
+	payment, ok := s.pendingPayments[paymentID]
+	if !ok {
+		payment = &pendingPayment{
+			ready: make(chan struct{}),
 		}
-
-		paymentErr = s.parseFailedPayment(payment, pkt, htlc)
-
-	default:
-		log.Warnf("Received unknown response type: %T", pkt.htlc)
-		return
+		s.pendingPayments[paymentID] = payment
 	}
+	s.pendingMutex.Unlock()
+
+	// Set the result packet and signal that to the callers it is ready.
+	payment.result = pkt
+	close(payment.ready)
```

Comment:
Should this happen before teardown? To prevent this:

```
Teardown circuit

        CommitCircuit (succeeds)

        get or create pending payment in memory (create)

get or create pending payment in memory + signal result

        async handoff to link if commit succeeded (pay twice)

        wait for pending payment result
```

Reply:
Yes, indeed! I also believe the pending payment should be created by `SendHTLC` after `CommitCircuit`, such that the two methods perform the operations in the opposite order: https://github.com/lightningnetwork/lnd/pull/2265/commits/6ffe4eb7183d09f96c4a2cec88472ec3dbd29873